Exchange Capabilities

RatkoR opened the pull request:
Mike, this is just a preview regarding the #746 issue.

I wanted to give you a head start on this and ask if it's still OK to do it. I feel it is, since every exchange now has it's own data inside it's own js file. But it's your call.

This patch series adds the `/exchanges` API which returns the same json as was in the `exchanges.js` file. It reads values from each exchange js file (by invoking [getCapabilities](https://github.com/RatkoR/gekko/blob/capabilities/exchanges/bitstamp.js#L163) method. UI works OK with it - see [here](https://github.com/RatkoR/gekko/blob/capabilities/web/vue/src/components/global/configbuilder/marketpicker.vue#L41).

There more work: fix all exchanges, fix exchangeChecker, etc... I'll do it if you're OK.

RatkoR committed:
API route that returns array of exchanges with their capabilities.

This is a new route. Up until now, exchange capabilities were hard
coded in /exchanges.js file. This patch series will move those
capabilities to the exchanges.

RatkoR committed:
Adds an API access point for the /exchanges route.

This commit is part of patch series where we move exchange
capabilities from one big /exchanges.js file to their
respective js files.

RatkoR committed:
Every exchange gets it's own capabilities method.

This commit is part of patch series where we move exchange
capabilities from one big /exchanges.js file to their
respective js files.

RatkoR committed:
Let marketpicker use the new API for getting exchange capabilities.

This commit is part of patch series where we move exchange
capabilities from one big /exchanges.js file to their
respective js files.

askmike commented:
Nice and clean implementation! Though it is limited to 3 exchanges. I'll leave this open for a while until I figure out what (I think) is best for Gekko :)

RatkoR commented:
No, no, this PR is not ready yet :) I added `getCapabilities` method to those three exchanges because I wanted to show you how an exchange would look like. And so that you can stop me before I loose too much time on it.

I'll change **all** exchanges before submitting this PR for real. I'll also fix `exchangeChaecker.js` and any other file that uses the `exchanges.js` file, since it is not going to be there any more.

askmike commented:
If you do all that & also update this documentation: https://gekko.wizb.it/docs/extending/add_an_exchange.html (by changing this file: https://github.com/askmike/gekko/blob/stable/docs/extending/add_an_exchange.md ) I have no choice but the accept :)

RatkoR committed:
Adds getCapabilities method to all exchanges.

Instead of having capabilities in a separate `exchamges.js` file,
each exchange has its own capabilities in its js file.

RatkoR committed:
Don't use `exchanges.js` file, since it's deprecated.

Exchange capabilities that were in `exchanges.js` file were
moved to their own .js files. For example, bitstamp's capabilities
are stored in bistamp.js file and can be retrieved by
`getCapabilities()` static method.

RatkoR committed:
Update documentation regarding exchange capabilities.

`exchanges.js` fiel is deprecated and we must update this doc
to reflect this change. Exchange capabilities are nov in its
own files. For example, bitstamp's capabilities are stored in
bistamp.js file and can be retrieved by `getCapabilities()`
static method.

RatkoR committed:
Remove not needed files.

`exchanges.js` and `markets.js` are deprecated since every
exchange hosts its own capabilities. For example, bitstamp's capabilities
are stored in bistamp.js file and can be retrieved by
`getCapabilities()` static method.

askmike commented on core/exchangeChecker.js:
```diff
@@ -12,10 +14,15 @@ Checker.prototype.notValid = function(conf) {
-  var exchange = _.find(exchanges, function(e) { return e.slug === slug });
+  var exchange = this.getExchangeCapabilities(slug);
 
   if(!exchange)
```
this if check won't get hit anymore, because [this line](https://github.com/askmike/gekko/pull/750/commits/02a62896e4d27adc2e04bc29f1382da9680461c7#diff-2a878ed84810bccfebe021d0e89d7ca8R18) will crash if the file is not found.

RatkoR replied:
True, and we don't want crashes - I'll add a try / catch block. Good catch!

RatkoR committed:
Bugfox: don't crash on wrong exchange name.

require() with a pathname that doesn't exists would crash
gekko so we need to guard against it.

RatkoR committed:
`exchanges.js` file was removed.

Exchange capabilities are fetched now by
`exchangeChecker.getExchangeCapabilities()` method.

RatkoR committed:
Use exchangeChecker object to get exchange capabilities since we need to instantiate it anyway.

RatkoR commented:
Allright Mike, I've tried as many ways to run gekko as I could: importer, ui, paper trading, npm tests - they all seemed to work OK. I didn't do realtime trading (I'm not trading real time with gekko yet). :)

I'd be glad if you can check the code and tell me what to change.

askmike commented:
Thanks a lot! Will try to check it out asap and let you know :)

askmike commented:
Sorry for the delay! Tested a bunch of features and everything seemed to work :) Will merge into develop and will do more thourough testing (together with #737) before it hits stable.
